src/utils/ShapeInformation.js

```diff
@@ -0,0 +1,215 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { getTypeFromName } from "../utils.js";
+import { Value, UndefinedValue, NullValue } from "../values/index.js";
+
+type AbstractValueType =
```

Comment:
And later you say that abstract values cannot in fact have `void` or `null` type.
Let's call this `ECMAScriptType`.